### PR TITLE
Resblock fix

### PR DESCRIPTION
**Pull request by AnnaAltynova**
fixed channels and groups in bottleneck case, refactored bottleneck to be separate Block

### CODE REVIEW

File: batchflow/models/torch/blocks/named_blocks.py [hunk 1]
```diff
@@ -120,23 +112,77 @@ def __init__(self, inputs=None, layout='cnacn', channels='same', kernel_size=3,
...
-                         'branch/stride': branch_stride_downsample}]
+                         'branch': branch_params}]
         layer_params += [{}]*(n_reps-1)
 
         super().__init__(*layer_params, inputs=inputs, layout=layout, channels=channels,
                          kernel_size=kernel_size, stride=stride, groups=groups, attention=attention,
                          **kwargs)
 
 
+class BottleneckBlock(ResBlock):
```
Comment (HollowPrincess) [inline batchflow/models/torch/blocks/named_blocks.py:131]: Why is it a separate class?

File: batchflow/models/torch/blocks/named_blocks.py [hunk 2]
```diff
@@ -120,23 +112,77 @@ def __init__(self, inputs=None, layout='cnacn', channels='same', kernel_size=3,
...
+    """ ResBlock with a canonical bottleneck (1x1 conv-batchnorm-activation) architecture.
```
Comment (HollowPrincess) [inline batchflow/models/torch/blocks/named_blocks.py:132]: `(1x1 conv-batchnorm-activation)` - is not the all truth, please, clarify.

File: batchflow/models/torch/blocks/named_blocks.py [hunk 3]
```diff
@@ -120,23 +112,77 @@ def __init__(self, inputs=None, layout='cnacn', channels='same', kernel_size=3,
...
+        If `str`, then number of channels is calculated by its evaluation. ``'S'`` and ``'same'`` stand for the
+        number of channels in the previous tensor. Note the `eval` usage under the hood.
+        If int, then number of channels in the output tensor. Default value is 'same'.
+    kernel_size : int, list of int
+        Convolution kernel size. Default is [1, 3, 1].
+    bottleneck : int
+        Factor of channels increase. Default is 4.
+    expand_channels : bool
+        If False, channels are increased in the first convolution layer only. Default is False.
+        If True, channels are increased in all but the last convolution layers.
```
Comment (HollowPrincess) [inline batchflow/models/torch/blocks/named_blocks.py:151]: `in all but the last` - ?

File: batchflow/models/torch/blocks/named_blocks.py [hunk 4]
```diff
@@ -120,23 +112,77 @@ def __init__(self, inputs=None, layout='cnacn', channels='same', kernel_size=3,
...
+    groups : int, list of int
+        Use `groups` convolution side by side, each  seeing 1 / `groups` the input channels,
```
Comment (HollowPrincess) [inline batchflow/models/torch/blocks/named_blocks.py:153]:
```suggestion
        Use `groups` convolution side by side, each seeing 1 / `groups` the input channels,
```

File: batchflow/models/torch/blocks/named_blocks.py [hunk 5]
```diff
@@ -120,23 +112,77 @@ def __init__(self, inputs=None, layout='cnacn', channels='same', kernel_size=3,
...
+        Example of input tensor to this layer.
+    layout : str
+        A sequence of letters, each letter meaning individual operation.
+        See more in :class:`~.layers.conv_block.BaseBlock` documentation. Default is 'cna cna cn'.
+    channels : int, str, list of int, list of str
+        If `str`, then number of channels is calculated by its evaluation. ``'S'`` and ``'same'`` stand for the
+        number of channels in the previous tensor. Note the `eval` usage under the hood.
+        If int, then number of channels in the output tensor. Default value is 'same'.
+    kernel_size : int, list of int
+        Convolution kernel size. Default is [1, 3, 1].
```
Comment (HollowPrincess) [inline batchflow/models/torch/blocks/named_blocks.py:146]:
```suggestion
        Convolution kernel size. Default is (1, 3, 1).
```

File: batchflow/models/torch/blocks/named_blocks.py [hunk 6]
```diff
@@ -120,23 +112,77 @@ def __init__(self, inputs=None, layout='cnacn', channels='same', kernel_size=3,
...
+
+        channels = [channels] * num_convs if isinstance(channels, (int, str)) else channels
+        channels = [safe_eval(item, get_num_channels(inputs)) if isinstance(item, str) else item
+                   for item in channels]
+        if expand_channels:
+            channels = [item * bottleneck for item in channels[:-1]] + [channels[-1]]
+        else:
+            channels[0] *= bottleneck
+
+        kernel_size = [kernel_size] * num_convs if isinstance(kernel_size, int) else list(kernel_size)
```
Comment (HollowPrincess) [inline batchflow/models/torch/blocks/named_blocks.py:175]: Why `list(kernel_size)`? Can we use the `kernel_size` instead of this?

File: batchflow/models/torch/blocks/named_blocks.py (hunk 1 above)
Comment (HollowPrincess) [reply batchflow/models/torch/blocks/named_blocks.py:131]: Also, we have much more files to replace `ResNet` with `bottleneck=True` option (e.g.: `densenet.py`, `resenet_with_stochastic_depth.py` , tutorial `00_project_from_scratch.ipynb`), please, don't forget about them.

File: batchflow/tests/notebooks/torch_test.ipynb [hunk 7]
```diff
@@ -1459,7 +1459,7 @@
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "Python 3 (ipykernel)",
+   "display_name": "Python 3",
```
Comment (HollowPrincess) [inline batchflow/tests/notebooks/torch_test.ipynb:1462]: Something gone wrong with this file update?

File: batchflow/models/torch/blocks/named_blocks.py [hunk 8]
```diff
@@ -120,23 +112,77 @@ def __init__(self, inputs=None, layout='cnacn', channels='same', kernel_size=3,
             # If main flow changes the number of channels, so must do the side branch.
             # No activation, because it will be applied after summation with the main flow
             branch_params = {'layout': branch_layout, 'channels': channels[-1],
-                             'kernel_size': 1, 'stride': branch_stride_downsample}
+                             'kernel_size': 1, 'stride': branch_stride_downsample, 'groups': branch_groups}
         else:
-            branch_params = {}
+            branch_params = {'stride': branch_stride_downsample, 'groups': branch_groups}
```
Comment (HollowPrincess) [inline batchflow/models/torch/blocks/named_blocks.py:117]: What's about providing `branch_layout` here too?

**Review by HollowPrincess — CHANGES_REQUESTED** [review]
Looks pretty nice.
I left some little comments. Please, don't forget to update other files with bottleneck ResBlocks.

File: batchflow/models/torch/blocks/named_blocks.py (hunk 1 above)
Comment (AnnaAltynova) [reply batchflow/models/torch/blocks/named_blocks.py:131]: I can fix the tutorial, but not sure about everithing else:
`densenet` uses `DenseBlock`, not the `ResBlock`, `resnet_with_stochastic_depth` is written with tensorflow and better needs his own PR, all the other `bottleneck`s are about attention

File: batchflow/models/torch/blocks/named_blocks.py (hunk 3 above)
Comment (AnnaAltynova) [reply batchflow/models/torch/blocks/named_blocks.py:151]: (except)

File: batchflow/models/torch/blocks/named_blocks.py (hunk 1 above)
Comment (HollowPrincess) [reply batchflow/models/torch/blocks/named_blocks.py:131]: You are right, only tutorial needs changes.

File: batchflow/models/torch/blocks/named_blocks.py (hunk 6 above)
Comment (AnnaAltynova) [reply batchflow/models/torch/blocks/named_blocks.py:176]: `Block` needs it as `list`

File: batchflow/models/torch/blocks/named_blocks.py (hunk 8 above)
Comment (AnnaAltynova) [reply batchflow/models/torch/blocks/named_blocks.py:118]: It would change default branch behaviour in case of no downsampling from `Identity()` to `cn`, but I can fix it

File: batchflow/tests/notebooks/torch_test.ipynb (hunk 7 above)
Comment (AnnaAltynova) [reply batchflow/tests/notebooks/torch_test.ipynb:1462]: no idea really

**Review by SergeyTsimfer — APPROVED** [review]
